packages/filesystem/src/node/download/file-download-handler.ts

```diff
@@ -43,46 +42,98 @@ export abstract class FileDownloadHandler {
+                // do not send filePath but instead use the downloadId
+                const data = { name, id: downloadId };
                 response.status(OK).send(data).end();
             });
         } catch (e) {
             this.handleError(response, e, INTERNAL_SERVER_ERROR);
         }
     }
 
+    protected async download(request: Request, response: Response, downloadInfo: DownloadStorageItem, id: string): Promise<void> {
+        const filePath = downloadInfo.file;
+        const statSize = downloadInfo.size;
+        // this sets the content-disposition and content-type automatically
+        response.attachment(filePath);
+        try {
+            await fs.access(filePath, fs.constants.R_OK);
+            response.setHeader('Accept-Ranges', 'bytes');
+            // parse range header and combine multiple ranges
+            const range = this.parseRangeHeader(request.headers['range'], statSize);
+            if (!range) {
+                response.setHeader('Content-Length', statSize);
+                this.streamDownload(OK, response, fs.createReadStream(filePath), id);
+            } else {
+                const rangeStart = range.start;
+                const rangeEnd = range.end;
+                if (rangeStart >= statSize || rangeEnd >= statSize) {
+                    response.setHeader('Content-Range', `bytes */${statSize}`);
+                    // Return the 416 'Requested Range Not Satisfiable'.
+                    response.status(REQUESTED_RANGE_NOT_SATISFIABLE).end();
+                    return;
+                }
+                response.setHeader('Content-Range', `bytes ${rangeStart}-${rangeEnd}/${statSize}`);
+                response.setHeader('Content-Length', rangeStart === rangeEnd ? 0 : (rangeEnd - rangeStart + 1));
+                response.setHeader('Cache-Control', 'no-cache');
+                this.streamDownload(PARTIAL_CONTENT, response, fs.createReadStream(filePath, { start: rangeStart, end: rangeEnd }), id);
+            }
+        } catch (e) {
+            this.fileDownloadCollection.deleteDownload(id);
+            this.handleError(response, e, INTERNAL_SERVER_ERROR);
+        }
+    }
+    protected streamDownload(status: number, response: Response, stream: fs.ReadStream, id: string) {
+        response.status(status);
+        stream.on('error', error => {
+            this.fileDownloadCollection.deleteDownload(id);
+            this.handleError(response, error, INTERNAL_SERVER_ERROR);
+        });
+        response.on('error', error => {
+            this.fileDownloadCollection.deleteDownload(id);
+            this.handleError(response, error, INTERNAL_SERVER_ERROR);
+        });
+        response.on('close', () => {
+            stream.destroy();
+        });
+        stream.pipe(response);
+    }
+    protected parseRangeHeader(range: string | string[] | undefined, statSize: number): { start: number, end: number } | undefined {
```

Comment:
Can we use [`range-parser`](https://www.npmjs.com/package/range-parser) instead?

Reply:
yes ofc, I'm just didn't want to add any other modules that are not used anywhere. But yeah, we can use that, do we have that on dependencies?

Reply:
It comes implicitly from `express`, but I would recommend declaring an explicit dependency with the same version (`1.2.0`) we have in the `yarn.lock`, so that we do not need a new CQ for it.

```
 yarn why range-parser
yarn why v1.15.2
[1/4] 🤔  Why do we have the module "range-parser"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "range-parser@1.2.0"
info Reasons this module exists
   - "_project_#@theia#core#express" depends on it
   - Hoisted from "_project_#@theia#core#express#range-parser"
   - Hoisted from "_project_#@theia#core#express#send#range-parser"
info Disk size without dependencies: "20KB"
info Disk size with unique dependencies: "20KB"
info Disk size with transitive dependencies: "20KB"
info Number of shared dependencies: 0
✨  Done in 0.90s.
```

If you do not have objections, I prefer such parsing maintained by others. `range-parser` has > 7M downloads per week.

Reply:
yeah no objection here. that was the original plan as well, I just didn't want to add any new module to avoid any cq work. :)

Reply:
@kittaakos I think its better just use `request.range(statSize)` its from [express ](https://expressjs.com/en/4x/api.html#req.range) too and i think it also uses `range-parser` in the background.